--- a/prosody.cfg.lua.dist
+++ b/prosody.cfg.lua.dist
@@ -53,6 +53,7 @@
 		"pep"; -- Enables users to publish their mood, activity, playing music and more
 		"register"; -- Allow users to register on this server using a client and change passwords
 		"adhoc"; -- Support for "ad-hoc commands" that can be executed with an XMPP client
+		--"smacks"; -- Support for XEP-0198
 
 	-- Admin interfaces
 		"admin_adhoc"; -- Allows administration via an XMPP client that supports ad-hoc commands
