--- a/usr/Makefile	2013-12-04 00:15:08.968385094 +0100
+++ b/usr/Makefile	2013-12-04 00:17:48.288765189 +0100
@@ -43,7 +43,7 @@
 
 LIBS += -lpthread -ldl
 
-PROGRAMS += tgtd tgtadm tgtimg
+PROGRAMS += tgtd tgtadm
 TGTD_OBJS += tgtd.o mgmt.o target.o scsi.o log.o driver.o util.o work.o \
 		concat_buf.o parser.o spc.o sbc.o mmc.o osd.o scc.o smc.o \
 		ssc.o libssc.o bs_rdwr.o bs_ssc.o \
@@ -70,14 +70,6 @@
 
 -include $(TGTADM_DEP)
 
-TGTIMG_OBJS = tgtimg.o libssc.o libcrc32c.o
-TGTIMG_DEP = $(TGTIMG_OBJS:.o=.d)
-
-tgtimg: $(TGTIMG_OBJS)
-	$(CC) $^ -o $@
-
--include $(TGTIMG_DEP)
-
 %.o: %.c
 	$(CC) -c $(CFLAGS) $*.c -o $*.o
 	@$(CC) -MM $(CFLAGS) -MF $*.d -MT $*.o $*.c
